Connection pool is not yet initialized

`pio import` in cluster mode is not working when I use mysql as event storage.
I think that it is related to connection pool of jdbc in singleton object.

```
pio import --appid 24 --channel cimport --input /tmp/unis/recommendation/317/events.json -- --driver-class-path /usr/share/java/mysql-connector-java.jar --master spark://ddsl0090:7077 --driver-memory 8G --executor-memory 10g --executor-cores 5 --total-executor-cores 40

[INFO] [Runner$] Submission command: /opt/spark//bin/spark-submit --conf spark.executor.extraClassPath=/usr/share/java/mysql-connector-java.jar --driver-class-path /usr/share/java/mysql-connector-java.jar --jars /usr/share/java/mysql-connector-java.jar --master spark://ddsl0090:7077 --driver-memory 8G --executor-memory 10g --executor-cores 5 --total-executor-cores 100 --class io.prediction.tools.imprt.FileToEvents --files file:/opt/pio/conf/log4j.properties,file:/etc/hadoop/conf/core-site.xml,file:/etc/hbase/conf/hbase-site.xml --driver-class-path /usr/share/java/mysql-connector-java.jar:/opt/pio/conf:/etc/hadoop/conf/:/etc/hbase/conf/ file:/opt/pio/lib/pio-assembly-0.9.5.jar --appid 24 --input /tmp/unis/recommendation/317/events.json --channel cimport --env PIO_ENV_LOADED=1,PIO_STORAGE_SOURCES_MYSQL_PASSWORD=pio,PIO_STORAGE_REPOSITORIES_METADATA_NAME=pio_meta,PIO_FS_BASEDIR=/opt/pio/.pio_store,PIO_STORAGE_SOURCES_MYSQL_URL=jdbc:mysql://ddsl0010/pio,PIO_HOME=/opt/pio,PIO_FS_ENGINESDIR=/opt/pio/.pio_store/engines,PIO_STORAGE_SOURCES_MYSQL_TYPE=jdbc,PIO_STORAGE_REPOSITORIES_METADATA_SOURCE=MYSQL,PIO_STORAGE_REPOSITORIES_MODELDATA_SOURCE=MYSQL,PIO_STORAGE_REPOSITORIES_EVENTDATA_NAME=pio_event,PIO_STORAGE_SOURCES_MYSQL_USERNAME=pio,PIO_FS_TMPDIR=/opt/pio/.pio_store/tmp,PIO_STORAGE_REPOSITORIES_MODELDATA_NAME=pio_model,PIO_STORAGE_REPOSITORIES_EVENTDATA_SOURCE=MYSQL,PIO_CONF_DIR=/opt/pio/conf
SLF4J: Class path contains multiple SLF4J bindings.
SLF4J: Found binding in [jar:file:/opt/spark-1.5.1-bin-hadoop2.6/lib/spark-assembly-1.5.1-hadoop2.6.0.jar!/org/slf4j/impl/StaticLoggerBinder.class]
SLF4J: Found binding in [jar:file:/opt/cloudera/parcels/CDH-5.4.7-1.cdh5.4.7.p0.3/jars/slf4j-log4j12-1.7.5.jar!/org/slf4j/impl/StaticLoggerBinder.class]
SLF4J: See http://www.slf4j.org/codes.html#multiple_bindings for an explanation.
SLF4J: Actual binding is of type [org.slf4j.impl.Log4jLoggerFactory]
[INFO] [Remoting] Starting remoting
[INFO] [Remoting] Remoting started; listening on addresses :[akka.tcp://sparkDriver@10.8.137.9:49801]
[WARN] [MetricsSystem] Using default name DAGScheduler for source because spark.app.id is not set.
[WARN] [DomainSocketFactory] The short-circuit local reads feature cannot be used because libhadoop cannot be loaded.
[Stage 0:>                                                                                                                                                                                                                                                          (0 + 2) / 2][WARN] [TaskSetManager] Lost task 0.0 in stage 0.0 (TID 0, 10.8.137.10): java.lang.IllegalStateException: Connection pool is not yet initialized.(name:'default)
	at scalikejdbc.ConnectionPool$$anonfun$get$1.apply(ConnectionPool.scala:58)
	at scalikejdbc.ConnectionPool$$anonfun$get$1.apply(ConnectionPool.scala:56)
	at scala.Option.getOrElse(Option.scala:120)
	at scalikejdbc.ConnectionPool$.get(ConnectionPool.scala:56)
	at scalikejdbc.ConnectionPool$.apply(ConnectionPool.scala:47)
	at scalikejdbc.DB$.connectionPool(DB.scala:150)
	at scalikejdbc.DB$.localTx(DB.scala:256)
	at io.prediction.data.storage.jdbc.JDBCPEvents$$anonfun$write$1.apply(JDBCPEvents.scala:141)
	at io.prediction.data.storage.jdbc.JDBCPEvents$$anonfun$write$1.apply(JDBCPEvents.scala:124)
	at org.apache.spark.rdd.RDD$$anonfun$foreachPartition$1$$anonfun$apply$29.apply(RDD.scala:898)
	at org.apache.spark.rdd.RDD$$anonfun$foreachPartition$1$$anonfun$apply$29.apply(RDD.scala:898)
	at org.apache.spark.SparkContext$$anonfun$runJob$5.apply(SparkContext.scala:1848)
	at org.apache.spark.SparkContext$$anonfun$runJob$5.apply(SparkContext.scala:1848)
	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:66)
	at org.apache.spark.scheduler.Task.run(Task.scala:88)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:214)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
	at java.lang.Thread.run(Thread.java:745)



[WARN] [TaskSetManager] Lost task 0.3 in stage 0.0 (TID 5, 10.8.137.12): java.lang.IllegalStateException: Connection pool is not yet initialized.(name:'default)
	at scalikejdbc.ConnectionPool$$anonfun$get$1.apply(ConnectionPool.scala:58)
	at scalikejdbc.ConnectionPool$$anonfun$get$1.apply(ConnectionPool.scala:56)
	at scala.Option.getOrElse(Option.scala:120)
	at scalikejdbc.ConnectionPool$.get(ConnectionPool.scala:56)
	at scalikejdbc.ConnectionPool$.apply(ConnectionPool.scala:47)
	at scalikejdbc.DB$.connectionPool(DB.scala:150)
	at scalikejdbc.DB$.localTx(DB.scala:256)
	at io.prediction.data.storage.jdbc.JDBCPEvents$$anonfun$write$1.apply(JDBCPEvents.scala:141)
	at io.prediction.data.storage.jdbc.JDBCPEvents$$anonfun$write$1.apply(JDBCPEvents.scala:124)
	at org.apache.spark.rdd.RDD$$anonfun$foreachPartition$1$$anonfun$apply$29.apply(RDD.scala:898)
	at org.apache.spark.rdd.RDD$$anonfun$foreachPartition$1$$anonfun$apply$29.apply(RDD.scala:898)
	at org.apache.spark.SparkContext$$anonfun$runJob$5.apply(SparkContext.scala:1848)
	at org.apache.spark.SparkContext$$anonfun$runJob$5.apply(SparkContext.scala:1848)
	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:66)
	at org.apache.spark.scheduler.Task.run(Task.scala:88)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:214)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
	at java.lang.Thread.run(Thread.java:745)

[ERROR] [TaskSetManager] Task 0 in stage 0.0 failed 4 times; aborting job
Exception in thread "main" org.apache.spark.SparkException: Job aborted due to stage failure: Task 0 in stage 0.0 failed 4 times, most recent failure: Lost task 0.3 in stage 0.0 (TID 5, 10.8.137.12): java.lang.IllegalStateException: Connection pool is not yet initialized.(name:'default)
	at scalikejdbc.ConnectionPool$$anonfun$get$1.apply(ConnectionPool.scala:58)
	at scalikejdbc.ConnectionPool$$anonfun$get$1.apply(ConnectionPool.scala:56)
	at scala.Option.getOrElse(Option.scala:120)
	at scalikejdbc.ConnectionPool$.get(ConnectionPool.scala:56)
	at scalikejdbc.ConnectionPool$.apply(ConnectionPool.scala:47)
	at scalikejdbc.DB$.connectionPool(DB.scala:150)
	at scalikejdbc.DB$.localTx(DB.scala:256)
	at io.prediction.data.storage.jdbc.JDBCPEvents$$anonfun$write$1.apply(JDBCPEvents.scala:141)
	at io.prediction.data.storage.jdbc.JDBCPEvents$$anonfun$write$1.apply(JDBCPEvents.scala:124)
	at org.apache.spark.rdd.RDD$$anonfun$foreachPartition$1$$anonfun$apply$29.apply(RDD.scala:898)
	at org.apache.spark.rdd.RDD$$anonfun$foreachPartition$1$$anonfun$apply$29.apply(RDD.scala:898)
	at org.apache.spark.SparkContext$$anonfun$runJob$5.apply(SparkContext.scala:1848)
	at org.apache.spark.SparkContext$$anonfun$runJob$5.apply(SparkContext.scala:1848)
	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:66)
	at org.apache.spark.scheduler.Task.run(Task.scala:88)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:214)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
	at java.lang.Thread.run(Thread.java:745)

Driver stacktrace:
	at org.apache.spark.scheduler.DAGScheduler.org$apache$spark$scheduler$DAGScheduler$$failJobAndIndependentStages(DAGScheduler.scala:1283)
	at org.apache.spark.scheduler.DAGScheduler$$anonfun$abortStage$1.apply(DAGScheduler.scala:1271)
	at org.apache.spark.scheduler.DAGScheduler$$anonfun$abortStage$1.apply(DAGScheduler.scala:1270)
	at scala.collection.mutable.ResizableArray$class.foreach(ResizableArray.scala:59)
	at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:47)
	at org.apache.spark.scheduler.DAGScheduler.abortStage(DAGScheduler.scala:1270)
	at org.apache.spark.scheduler.DAGScheduler$$anonfun$handleTaskSetFailed$1.apply(DAGScheduler.scala:697)
	at org.apache.spark.scheduler.DAGScheduler$$anonfun$handleTaskSetFailed$1.apply(DAGScheduler.scala:697)
	at scala.Option.foreach(Option.scala:236)
	at org.apache.spark.scheduler.DAGScheduler.handleTaskSetFailed(DAGScheduler.scala:697)
	at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.doOnReceive(DAGScheduler.scala:1496)
	at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:1458)
	at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:1447)
	at org.apache.spark.util.EventLoop$$anon$1.run(EventLoop.scala:48)
	at org.apache.spark.scheduler.DAGScheduler.runJob(DAGScheduler.scala:567)
	at org.apache.spark.SparkContext.runJob(SparkContext.scala:1822)
	at org.apache.spark.SparkContext.runJob(SparkContext.scala:1835)
	at org.apache.spark.SparkContext.runJob(SparkContext.scala:1848)
	at org.apache.spark.SparkContext.runJob(SparkContext.scala:1919)
	at org.apache.spark.rdd.RDD$$anonfun$foreachPartition$1.apply(RDD.scala:898)
	at org.apache.spark.rdd.RDD$$anonfun$foreachPartition$1.apply(RDD.scala:896)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:147)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:108)
	at org.apache.spark.rdd.RDD.withScope(RDD.scala:306)
	at org.apache.spark.rdd.RDD.foreachPartition(RDD.scala:896)
	at io.prediction.data.storage.jdbc.JDBCPEvents.write(JDBCPEvents.scala:124)
	at io.prediction.tools.imprt.FileToEvents$$anonfun$main$1.apply(FileToEvents.scala:90)
	at io.prediction.tools.imprt.FileToEvents$$anonfun$main$1.apply(FileToEvents.scala:63)
	at scala.Option.map(Option.scala:145)
	at io.prediction.tools.imprt.FileToEvents$.main(FileToEvents.scala:63)
	at io.prediction.tools.imprt.FileToEvents.main(FileToEvents.scala)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at org.apache.spark.deploy.SparkSubmit$.org$apache$spark$deploy$SparkSubmit$$runMain(SparkSubmit.scala:672)
	at org.apache.spark.deploy.SparkSubmit$.doRunMain$1(SparkSubmit.scala:180)
	at org.apache.spark.deploy.SparkSubmit$.submit(SparkSubmit.scala:205)
	at org.apache.spark.deploy.SparkSubmit$.main(SparkSubmit.scala:120)
	at org.apache.spark.deploy.SparkSubmit.main(SparkSubmit.scala)
Caused by: java.lang.IllegalStateException: Connection pool is not yet initialized.(name:'default)
	at scalikejdbc.ConnectionPool$$anonfun$get$1.apply(ConnectionPool.scala:58)
	at scalikejdbc.ConnectionPool$$anonfun$get$1.apply(ConnectionPool.scala:56)
	at scala.Option.getOrElse(Option.scala:120)
	at scalikejdbc.ConnectionPool$.get(ConnectionPool.scala:56)
	at scalikejdbc.ConnectionPool$.apply(ConnectionPool.scala:47)
	at scalikejdbc.DB$.connectionPool(DB.scala:150)
	at scalikejdbc.DB$.localTx(DB.scala:256)
	at io.prediction.data.storage.jdbc.JDBCPEvents$$anonfun$write$1.apply(JDBCPEvents.scala:141)
	at io.prediction.data.storage.jdbc.JDBCPEvents$$anonfun$write$1.apply(JDBCPEvents.scala:124)
	at org.apache.spark.rdd.RDD$$anonfun$foreachPartition$1$$anonfun$apply$29.apply(RDD.scala:898)
	at org.apache.spark.rdd.RDD$$anonfun$foreachPartition$1$$anonfun$apply$29.apply(RDD.scala:898)
	at org.apache.spark.SparkContext$$anonfun$runJob$5.apply(SparkContext.scala:1848)
	at org.apache.spark.SparkContext$$anonfun$runJob$5.apply(SparkContext.scala:1848)
	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:66)
	at org.apache.spark.scheduler.Task.run(Task.scala:88)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:214)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
	at java.lang.Thread.run(Thread.java:745)
[s.chon@ddsl0090 deep-recommender-unis-script]$
```

To solve this problem, I run it in local mode like this.
```pio import --appid 24 --channel cimport --input /tmp/unis/recommendation/317/events.json -- --driver-class-path /usr/share/java/mysql-connector-java.jar --master local[8]```

That works find.

In my opinion, We should upgrade to spark version 1.4.0 or higher to take advantage of good method of spark data frame.

I tried to do this. It works find in cluster mode.

--- a/build.sbt
+++ b/build.sbt
@@ -37,7 +37,7 @@ elasticsearchVersion in ThisBuild := "1.4.4"
 
 json4sVersion in ThisBuild := "3.2.10"
 
-sparkVersion in ThisBuild := "1.3.0"
+sparkVersion in ThisBuild := "1.4.0"
 
 lazy val pioBuildInfoSettings = buildInfoSettings ++ Seq(
   sourceGenerators in Compile <+= buildInfo,

--- a/data/src/main/scala/io/prediction/data/storage/jdbc/JDBCPEvents.scala
+++ b/data/src/main/scala/io/prediction/data/storage/jdbc/JDBCPEvents.scala
@@ -15,20 +15,15 @@
 
 package io.prediction.data.storage.jdbc
 
-import java.sql.DriverManager
-import java.sql.ResultSet
+import java.sql.{DriverManager, ResultSet}
 
 import com.github.nscala_time.time.Imports._
-import io.prediction.data.storage.DataMap
-import io.prediction.data.storage.Event
-import io.prediction.data.storage.PEvents
-import io.prediction.data.storage.StorageClientConfig
+import io.prediction.data.storage.{DataMap, Event, PEvents, StorageClientConfig}
 import org.apache.spark.SparkContext
-import org.apache.spark.rdd.JdbcRDD
-import org.apache.spark.rdd.RDD
+import org.apache.spark.rdd.{JdbcRDD, RDD}
+import org.apache.spark.sql.{SQLContext, SaveMode}
 import org.json4s.JObject
 import org.json4s.native.Serialization
-import scalikejdbc._
 
 /** JDBC implementation of [[PEvents]] */
 class JDBCPEvents(client: String, config: StorageClientConfig, namespace: String) extends PEvents {
@@ -119,32 +114,47 @@ class JDBCPEvents(client: String, config: StorageClientConfig, namespace: String
   }
 
   def write(events: RDD[Event], appId: Int, channelId: Option[Int])(sc: SparkContext): Unit = {
-    @transient lazy val tableName =
-      sqls.createUnsafely(JDBCUtils.eventTableName(namespace, appId, channelId))
-    events.foreachPartition { events =>
-      val batchParams = events.map { event =>
-        Seq(
-          event.eventId.getOrElse(JDBCUtils.generateId),
-          event.event,
-          event.entityType,
-          event.entityId,
-          event.targetEntityType,
-          event.targetEntityId,
-          Serialization.write(event.properties.toJObject),
-          event.eventTime,
-          event.eventTime.getZone.getID,
-          if (event.tags.nonEmpty) Some(event.tags.mkString(",")) else None,
-          event.prId,
-          event.creationTime,
-          event.creationTime.getZone.getID)
-      }.toSeq
-      DB localTx { implicit session =>
-        @transient lazy val q =
-          sql"""
-          insert into $tableName values(?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
-          """
-        q.batch(batchParams: _*).apply()
-      }
-    }
+    val sqlContext = new SQLContext(sc)
+
+    import sqlContext.implicits._
+
+    val tableName = JDBCUtils.eventTableName(namespace, appId, channelId)
+
+    val eventTableColumns = Seq[String](
+        "id"
+      , "event"
+      , "entityType"
+      , "entityId"
+      , "targetEntityType"
+      , "targetEntityId"
+      , "properties"
+      , "eventTime"
+      , "eventTimeZone"
+      , "tags"
+      , "prId"
+      , "creationTime"
+      , "creationTimeZone")
+
+    val eventDF = events.map { event =>
+      (event.eventId.getOrElse(JDBCUtils.generateId)
+        , event.event
+        , event.entityType
+        , event.entityId
+        , event.targetEntityType.orNull
+        , event.targetEntityId.orNull
+        , if (!event.properties.isEmpty) Serialization.write(event.properties.toJObject) else null
+        , new java.sql.Timestamp(event.eventTime.getMillis)
+        , event.eventTime.getZone.getID
+        , if (event.tags.nonEmpty) Some(event.tags.mkString(",")) else null
+        , event.prId
+        , new java.sql.Timestamp(event.creationTime.getMillis)
+        , event.creationTime.getZone.getID)
+    }.toDF(eventTableColumns:_*)
+
+    // spark version 1.4.0 or higher
+    val prop = new java.util.Properties
+    prop.setProperty("user", config.properties("USERNAME"))
+    prop.setProperty("password", config.properties("PASSWORD"))
+    eventDF.write.mode(SaveMode.Append).jdbc(client, tableName, prop)
   }
 }

